std/assembly/util/uri.ts

```diff
@@ -0,0 +1,260 @@
+import { E_URI_MALFORMED } from "./error";
+import { CharCode } from "./string";
+
+// @ts-ignore: decorator
+@lazy export const URI_UNSAFE = memory.data<u8>([
+/* skip 32 + 1 always set to '1' head slots
+  */ 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
```

Comment:
These still feels like a case for a bitmap to me, saving ~7/8 bits.

Reply:
binaryen pretty great pack this. all zero ranges folded so such LUTs really small actually. Just little bit larger (10-25 bytes) than bit-packed u32/u64 values in my tests. But have better performance and could easily customized (if spec changed for example)

Reply:
See this comment: https://github.com/AssemblyScript/assemblyscript/pull/1733#discussion_r602319126

So bit-packet code takes: 118 bytes
optimized table (LUT): 126 bytes

Reply:
Would it be possible to optimize the bit-packed table in a similar way?

Reply:
Hmm, I'm a little confused. The `u8` table has `16 * 8 - 1 = 127` bytes (1016 bits) while the bit-packed one has `4 * 4 = 16` bytes (128 bits). What am I missing? :)

Reply:
yeah that's true. But if LUT is qute sparse (has a lot of zeros) it greatly compressed by binaryen. So need only one bit cell need into account =) Also bit-packed u32/u64 set required unpacking which much more complicated and add extra bytes to total size

Reply:
And of course if lut table will take >= 512 I guess bit-packed will be more prefereable. But for such small tables benefit very small (less 10 bytes)

Reply:
Good argument regarding code size, yeah. I guess I am fine with trading a few bytes in memory for binary size.

Reply:
Looking at the code again, we are talking about

```ts
  if ((ch - 33) < 0x7F) {
    return <bool>load<u8>(URI_SAFE_OPT_2 + (ch - 33));
  }
```

vs

```ts
  if (ch < 0x7F) {
    return <bool>((load<u32>(URI_SAFE_PACKED + (((ch >>> 5) >>> ~ch)) << 2)) & 1);
  }
```

essentially, right?

Reply:
yeah)

Reply:
Actually `ch < 0x7F` unnecessary for `URI_SAFE_PACKED` but it doesn't significantly change whole picture

Reply:
Overall that doesn't look too bad, though, given that memory (at runtime) is significantly less, and that code size only increases by a few bytes. I'd probably go for it, and annotate the u32s as binary strings to indicate what's going on, hmm...

Reply:
even with slower execution? I prefer current implementation because it's little bit less hacky (although this is of course pretty relative =)) and definitely faster.

Reply:
Hmm, additional loads would be bad, but aren't a few bitwise ops or such in a row essentially free anyway?

Reply:
original check
```wat
    get_local $p0
    i32.const 33
    i32.sub
    i32.const 127
    i32.lt_u
    if $I0
      get_local $p0
      i32.load8_u offset=991
      i32.const 0
      i32.ne
      return
    end
    i32.const 0
```
pit-packed check:
```wat
    get_local $p0
    i32.const 127
    i32.lt_u
    if $I0
      get_local $p0
      i32.const 5
      i32.shr_u
      get_local $p0
      i32.const -1
      i32.xor
      i32.shr_u
      i32.const 1120
      i32.add
      i32.const 2
      i32.shl
      i32.load
      i32.const 1
      i32.and
      return
    end
    i32.const 0)
```

Reply:
Well, I don't know for sure how pipeline friendly this is on modern CPUs, but I have a feeling that this isn't measurably slower and will mostly boil down to the time the `load` takes. Dangerous half-knowledge, though.

Reply:
I'm too lazy to make benchmark so created C++ version and measured via llvm-mca:
original:
```
Iterations:        100
Instructions:      800
Total Cycles:      245
Total uOps:        1300

Dispatch Width:    6
uOps Per Cycle:    5.31
IPC:               3.27
Block RThroughput: 2.2
```

bit-packed:
```
Iterations:        100
Instructions:      1200
Total Cycles:      379
Total uOps:        1800

Dispatch Width:    6
uOps Per Cycle:    4.75
IPC:               3.17
Block RThroughput: 3.0
```

fiddle: https://godbolt.org/z/daGceacW8

So as you can see second variand ~1.5x time should be slower

Reply:
I see, thanks for checking :) I don't know enough about this I fear, and since it isn't super critical, I am OK with doing it as you suggest.